### ISSUE TITLE
Is Parse-SDK compatible with .NET Standard ?

### ISSUE THREAD

**Issue by Binnette** [parse-community/Parse-SDK-dotNET#274]
Hi,

Is Parse-SDK-dotNET compatible with .Net Standard ?
If yes, which versions of .Net Standard are supported ?

Also, is there any ".NET Standard" instructions ?

For now, i install "Parse 1.7.0", but I get this NuGet warning is VS 2017:

> Warning NU1701: Package 'Parse 1.7.0' was restored using '.NETFramework,Version=v4.6.1' instead of the project target framework '.NETStandard,Version=v2.0'. This package may not be fully compatible with your project.

Thank you

**Comment by flovilmart** [comment]
https://github.com/dotnet/standard/blob/master/docs/netstandard-20/README.md#net-framework-461-supporting-net-standard-20

Maybe we can move completely to .NetStandard if we don’t use one of the 43 api’s framework provides and standard don’t

**Comment by Binnette** [comment]
In fact there are 52 (and not 43) unsupported API, see dotnet/standard#524

Here are the 52 unsupported API:
https://github.com/dotnet/standard/blob/master/netstandard/src/ApiCompatBaseline.net461.txt

Can anyone please check in ParseSDK if any of those API are used ?

**Comment by flovilmart** [comment]
Could you check for us? We’re a little low on .Net experts :)

**Comment by montymxb** [comment]
Yeah we're a bit short on that end at the moment :/. Any insight you can provide in that regard would be greatly appreciated. In December I expect to be able to sit down and start some work on this sdk. It's long overdue for it!

**Comment by Binnette** [comment]
@flovilmart : i do not find any usage of those 52 unsupported API.

**Comment by flovilmart** [comment]
So in practice that means we can safely target the expected library ? What would be the next steps?

**Comment by flovilmart** [comment]
Seems that we can silence the warning, https://stackoverflow.com/questions/45738982/how-to-hide-net-core-compatibility-warnings-when-referencing-net-4-6-nuget-pac not sure what it does in the end or what steps are to be taken.

**Comment by montymxb** [comment]
Maybe we should just bump up .NET standard. Would be nice to catch up after this recent period of inactivity.

**Comment by TheFanatr** [comment]
I rewrote the entire thing into one netstandard2.0 assembly as a hobby completely separately. It works fine but I changed a couple of things because it wasn't meant to be used officially as a replacement to this version. I also got pissed off at Unity and Mono so I deleted all of that stuff, and I also got rid of the unit tests. Oh well; I can add those back. The original point was that I wanted to add Parse LiveQuery, then I went off on a tangent and converted everything into .NET Standard. I may or may not have too much free time. Anyways, I may make a public version at some point soon; right now the project is hosted on a private repository on my friend's account because we were planning to do some stuff with it. Let me know if any of you want to have a poke. Thanks,

**Comment by TheFanatr** [comment]
PCLStorage may be an issue. I recently tried to reimplement the unit tests with a .NET Core unit test project, and 45 of the tests failed due to the fact that PCLStorage uses System.Windows.Forms in order to get the temp path when used on Windows platforms. I dug around through referencesource.microsoft.com and the functionality can easily be replicated without having to use the assembly; however, that would mean writing a separate version of PCLStorage and/or submitting a pull request. The first option may be tricky with getting people to remember to reference the modified version when they use parse. The second option may take a lot of time depending on the person who manages the PCLStorage GitHub project. My other thought was to have platform-specific Parse.Storage.TARGET_PLATFORM_HERE projects that implement storage, that need to be referenced when parse is in use on specific platforms. Alternatively, to simplify that, it is also possible to do what PCLStorage does by having one project and use Roselyn "preprocessor directives" to make the separate platforms work. A project was started by someone else that aims to do a similar thing, only that the platform detection method used is programmatic which may or may not cause issues. Is there a simpler solution that I am missing?

**Comment by montymxb** [comment]
@TheFanatr It looks like PCLStorage is [not being actively maintained](https://github.com/dsplaisted/PCLStorage/issues/66) (was looking at this earlier today in relation to some other issues). We're probably going to want to go towards setting up an internal ParseStorage subproject. No exact directions on this yet, and we're definitely open to completely new implementations (or decently supported 3rd party equivalents).

Just to keep you in the loop we're in the midst of a proto-release, probably something you could consider an alpha more or less. I've personally put together Core, Common and some other parts, but haven't had too much success getting things hooked backed up with Unity, Mono, etc. We'll be moving this along with a release containing dlls and updated build instructions. From there we'll start addressing issues and move towards fully supporting other platforms (Unity, Xamarin, etc.), with the possibility of going .NET standard while we're at it. 

Baby steps essentially, but we're looking to start making quite a few of them pretty soon here. If you have any thoughts as we start to do this (or would like to put some PRs up) be sure to let us know. This is going to be a WIP for a bit so we'll be around here.

**Comment by TheFanatr** [comment]
I see. Yeah, I basically refactored the entirety of all the Parse projects into one assembly (one project that targets .NET Standard simply called Parse), so a pull request is probably out of the question. I kept the entire API around except ParseAnatlytics and ParsePush because I originally was just going to use this with a friend. I will soon try to bring ParseAnalytics back in, but ParsePush is way too much of a pain so I came up with an alternative that I did not yet implement. Basically, the idea was to take advantage of C# events to set up a push notification hooking system, so that people can do whatever they want when the notification arrives. This would also alleviate the need to hold ParsePush in a whole bunch of separate assemblies. What do you think? If the issue of being able to have the notification detector running in the background needs to be solved, it is possible to use Rx.Net's ability to run specified delegates in the background very efficiently on a separate thread. Anyways, I will probably not deal with that until after I solve the cross-platform storage problem.

As for that, what could be done is take advantage of the fact that you don't need to reference an assembly that a certain library references, as long as you don't encounter code that needs it to be there. What this means is that the Parse project can have references to Xamarin, and UWP assemblies and use them to deal with storage, but include preprocessor directives to exclude their usage in code when Parse is being used on platforms where those assemblies are not needed. This way, for example, someone using Parse from an iOS will have no issue using the storage features because the Xamarin assembly required will already be referenced by their project. The only issue I see with this is getting an FileNotFoundException when targeting an assembly version other than the one referenced in the Parse project. The only solution to that that I see uses reflection to detect whether the needed namespaces, classes and methods are present. Do I have this correct?

**Comment by TheFanatr** [comment]
As for suggestions for the proto-release, it would be great to see a de-cluttering of the project; unsupported-framework-based projects should dissapear, and the Unity projects should be unified (as it were) so that the functionality isn't so smashed apart, and an upgrade to the latest supported language version and .NET implementation by Unity for the Unity projects. I think that the latest .NET implementation supported by unity is .NET Framework 4.6, and the latest version of the C# language spec supported by unity is version 6.

**Comment by montymxb** [comment]
1. Cross Platform Storage (PCL step up, assuming PCL does not work??)
PCL is working for now, but it is a bit weird that it's not clarified as an external dependency. We'll be clarifying it's necessity in future build instructions. With regard to any shortcomings we'll want to start on an integrated solution. The existing test suite is functional as is and we'll be keeping that, if anything we'll be expanding on it and potentially looking towards a functional test suite as well against a mock server (similar to what we do for the php sdk).

I'm not entirely certain but I believe reflection is being used internally for some processes. This is a tough one to weigh in on, as this is an 'inherited' project that was in mid-work. We've been working from the ground up to determine where we stand and where we need to go from here.

Despite the mid-work note the project is well organized. Unfortunately it's not documented in regards to it's new structure, as the original maintainers were in the process of modularizing the sdk. The intent was to decouple the independent modules, allowing them to be used independently of each other (push, analytics, etc.). In the interest of preserving the original vision we'll be continuing with this modularization, but our intent is to provide the original **Parse.dll** that is all encompassing. The new modules would be available as well, providing the ability for users to pick and choose components rather than the 'super' solution.

It goes without saying that one of our priorities is to provide updated build & implementation instructions so that others can start working with this new format pronto.

The unity integrations need some looking into. I've built unity dlls but have had little luck with a successful integration, same goes for Xamarin/Mono. Whatever we put together for unity will be following the modular convention as mentioned above. I can provide more details once we've gotten into Unity related integrations.

**Comment by TheFanatr** [comment]
> The intent was to decouple the independent modules, allowing them to be used independently of each other (push, analytics, etc.). In the interest of preserving the original vision we'll be continuing with this modularization, but our intent is to provide the original Parse.dll that is all encompassing.

I understand why the ability to pick and choose which sections of the SDK seems like a good idea, but at the end of the day, what benefit does it really provide? It will only make managing the project harder, and make the structure more convoluted. In my opinion, there is no real-world reason to prefer referencing an assembly that only provides a portion of a certain SDK, rather than the whole thing, unless it is extremely massive. The extra file size is negligible with today's devices. It would greatly simplify everything. Is there a reasoning? I mean, if this kind of thing really mattered that much, why is mscorlib, which contains a ton of unrelated things, one assembly? To clarify, by unrelated I mean varied and not strictly connected, and also when I say mscorlib I am referring to the mscorlib.dll assembly that comes with the latest version of .NET Framework.

As for the unit test thing I mentioned in a previous comment, the reason I brought it up was because I have been having trouble properly using Parse from pure .NET Core in certain cases, mainly due to PCLStorage having questionable methodology for creating storage directories (relying on System.Windows.Forms). This was especially apparent in the unit test. So my point was that relying on a project that runs on outdated technology (PCLs), where the functionality can now be completely implemented in .NET Standard, doesn't seem logical to me especially when considering that it doesn't work in certain cases and now it's no longer being actively developed.

Also, just a question: does this planned modularization mean that you are ditching .NET Standard in preference for .NET-implementation-specific projects?

**Comment by montymxb** [comment]
@TheFanatr the benefit is to allow independent use of components, without the whole SDK. If you want to open a discussion regarding that specifically can you open another thread (this one should be focused on .NET standard after all).

We also are aware about the concern regard PCLs being out of date, and we are planning to make a move to .NET standard. Just to clarify our focus is as follows (and in order ideally):
- ~Setup CI~ (done)
- Get working builds for the current setup
- Present a working release for the current setup
- With working build correct existing issues and take community feedback
- Move towards standard

We're not trying to bypass .NET standard or anything like that, my apologies if we're giving that impression. Current priorities place it _last_ considering we haven't had a working release in some time and there are quite a few issues that have arisen since then. Some of them may need to be corrected by switching to standard, but first and foremost we will present a workable release that we can then base what we need to change off of. 

This means continuing with the modularization since it is the current direction of the sdk. But I should clarify the current team, myself included, did _not_ set it up this way intentionally. We're simply seeing through the plans of the original maintainers, who are no longer available to work on this project. It's quite possible that we may ditch the modularization if it becomes either too problematic or unused (or both) and is simply not worth the structure. If we do ditch and revert to a singular design, I would like to knowing that we are doing so from a previously stable build.

Thanks for taking the time to present us your thoughts on this by the way, we actually really like seeing that others are interested in the sdks enough to push for changes 👍 . Just give us some time if it seems we're moving slow, we like to make our moves carefully, albeit not always perfectly in some cases.

**Comment by flovilmart** [comment]
I actually believe modularisation will help moving to .Net standard as an additional target. I dunno if it makes sense in 2018, but on iOS we still target iOS 8 as a minimum platform. We could very well undo the modularization commits and get back to the original stage. I'm not sure what's best now. @baywet what do you think?

**Comment by TheFanatr** [comment]
@montymxb @flovilmart So, I have an idea that will both allow modularity, make the project much simpler, and be open enough to easily change the structure later on. It's actually quite simple: create multiple .NET Standard assemblies with all the modules that are wanted to be separated. Literally all of the PCL code works out of the box via copy and paste except a few minor required changes, and the UWP AssemblyLister works perfectly on .NET Standard. I actually have a fully working implementation of the Parse SDK, minus Analytics and Push all in one assembly, with unit tests and all.
![image](https://user-images.githubusercontent.com/16641853/34505262-0bb0efb0-efd9-11e7-877b-875b829b6cca.png)
All I had to do was rewrite PCLStorage into .NET Standard to make the unit tests work and alleviate the need to use "framework libraries" such as Windows Forms.
![image](https://user-images.githubusercontent.com/16641853/34506168-4d667c58-efdf-11e7-85ed-7a965b17edf1.png)
It's called StandardStorage and it keeps 100% of the public PCLStorage API, in terms of both functionality and usage, only that it's in a different namespace. I may create some repositories to hold the projects at some point hopefully soon. I can also try to modularize my version of Parse, if you would like. My next step with Parse it getting Analytics back in, then I will go ahead with the modularization; again if you guys want. The point is that it's already possible to convert the entire API into one assembly so modularization doesn't actually help with anything except change the structure of the project. So really, now it's up to a project design perspective; will it actually be useful to people to have only one module in their code (ex.: only using ParsePush, but not the core functionality such as querying and ParseObject creation)?

**Comment by flovilmart** [comment]
@TheFanatr 

> So really, now it's up to a project design perspective; will it actually be useful to people to have only one module in their code (ex.: only using ParsePush, but not the core functionality such as querying and ParseObject creation)?

I belive it was intented the other way around, being able to use Core without Push as we don't maintain a push adapter for windows targets. 

Analytics don't really matter as it's 100% noops serverside, unless you provide your own adapter so that's interesting to keep it separated to keep a lower binary size.

In the end, the way the project is architectured and distributed doesn't really matter, what's the most important is the developer experience both in consuming and developing the project.

Would you be willing to do a PR on this repo and become a maintainer of it?

**Comment by TheFanatr** [comment]
@flovilmart Sure! I will just have to actually fork as I was developing this completely separately. Also, considering I am changing so much about the structure it is probably impossible to merge without complete erasure of the current code, or is it possible to overwrite? Either way, I will probably post it sometime within the next 5 hours or so.

**Comment by TheFanatr** [comment]
I will start by putting StandardStorage into it's own repository, and creating a NuGet package for it and reference it from Parse, then I will submit a PR with the repo in it's current state.

**Comment by baywet** [comment]
@flovilmart usually on big projects have one nuget with multiple DLL's, it gives more flexibility than only one dll to consumers if they want to reduce the size of their application (by referencing only specific dlls).
If i remember properly, by default nugets with multiple dll will reference all dlls in project they are being installed during installation time. And I'd say not a lot of people care about the size of their app besides those doing IoT or Mobile dev to the point they go ahead and de-reference some dlls.  
A better approach is to modularize in multiple nugets. I brings the flexibility I was mentioning above with a few other perks:
- dedicated doc per nuget
- better understanding of the intent (naming conventions...)
- dependency management (let's say that parse needs nuget A and B and Parse.Something A B and C, if somebody only needs parse, you don't want to force them to install C as well by having an only package)
- most of the organizations including Microsoft went this way over the last couple of years (look are azure storage, iot hub, webjobs...)

Now I understand you already have multiple packages. **First it seems they are not all owned by the Parse account on nuget, which might cause you issues, you probably want to take back ownership** if they come from that repo. Then, **I don't think it's good to put everything back together in one package mostly because it will mean a big breaking change**, people will have to work around their dependencies, clean things etc.

All that being said and talking about the .net standard, just to make sure everybody is on the same page: **the lower version of the standard you implement, the more "systems" you will support but the harder it will be for you** (less API's available). I suggest you give a close look at [this page](https://docs.microsoft.com/en-us/dotnet/standard/net-standard) and if you can back it up by telemetry to see what platforms your community is developing against.

**Comment by TheFanatr** [comment]
> let's say that parse needs nuget A and B and Parse.Something A B and C, if somebody only needs parse, you don't want to force them to install C

@baywet Just saying, Parse doesn't reference any NuGet packages except PCLStorage, so I don't see how your statement currently applies; however, with a hypothetical situation in which Parse did reference multiple packages, it could possibly be useful to have multiple NuGets, I agree. 

@flovilmart Since this SDK still doesn't a ParseLiveQuery implementation, I can see how it is possible that some packages could be needed in the future, so I think I will try to go ahead with modularizing my code base, after submitting the pull request, for the sake of users being able to micro-optimize their dependencies.

# 19000 New APIs in .NET Standard 2 versus 1.6
As for that .NET Standard implementation support chart, I know it would be a breaking change, but it's within reason. I have had to use that chart before for making libraries, but I usually go against the lowest-common-denominator approach, in terms of platform-specifc portability. The reason is that actual maintainability is much more important. Right now, the entire project is ziptied to the rotting support beams of deprecated technology, and there's with a better alternative. PCLs rely on API intersections in between specific platforms, whereas .NET Standard is an API inclusion standard that forces platforms to comply in order to achieve support. What's important about that is that .NET Standard is universal, whereas PCLs are specific to profile. By converting to .NET Standard, any platform that implements version 2 into their API will be able to use Parse if we build it targeting the latest version of the standard, meaning that we would be supporting more potential platforms off the bat, but loosing support for very old platforms that are no longer being updated. Basically the tradeoff is that we loose certain (no-longer-in-development and/or unsupported platforms), but gain the ability for the package to be used universally. Getting back to development ease: by targeting version 2, support for older .NET Core and Xamarin platforms is lost, but the project gains access to **19 000 new APIs over the previous version**! This means that, for the platforms that do implement/support the latest version, much more is possible. The speed at which features can be implemented and new technologies can be taken advantage of is much higher than if the lowest-common-denominator route is taken.

**Comment by baywet** [comment]
@thefanatr : I agree, I was only providing insights about how things are generally done in DotNet and I have little knowledge of this project.
If you want to keep PCL support and old platforms with it you could either rely on conditional build with multiple build profiles or on conditional build with multiple project files. And then "manually" package your nugget. (Instead of relying on default DotNet pack Command).
That's going to require a lot of works though to get things set up and increase maintenance cost over time.
I'd advocate for a new major version (or many if the direction of multiple packages is taken) that brings support for core and tells people to upgrade their projects if they want the new stuff.
You can always start a support branch for this version.
I hope all my comments helped the thought process a bit :)

**Comment by TheFanatr** [comment]
@baywet Yes, I think a new major version would be in order. As for using conditional builds, I think that approach would be very time-consuming, but could work, if needed; good suggestion. 

@flovilmart I will try to get the PR posted soon; I was just performing some maintenance tasks to get it ready, and I will probably try to get at least analytics working.

Also, StandardStorage is on NuGet; it's my .NET Standard 2 reimplementation of PCLStorage that is 100% compatible with all .NET Standard 2 platforms that have equivalents for AppData and LocalAppData from the Environment.SpecialPath enum. Also it has all of the portable public API from PCLStorage, so it's literally an in-place no-modification replacement.

On a side note, does anyone have any idea on how it would be possible to still support Unity without having to link every file into a Unity scripting project and use the preprocessor (conditionals)?

**Comment by TheFanatr** [comment]
Do I need to make a portable equivalent to TrackAppOpenedAsync with the platform specific launch args as a parameter? Such method exists on both supported mobile platforms (WinRT and Windows Phone). The only thing it really does as far as I can tell is try to extract the push hash, if existing, from application launch metadata; is there a special purpose this serves? There is already a TrackAppOpenedWithPushHashAsync method, so I assumed that the answer is no, but thought to check just in case.

**Comment by flovilmart** [comment]
I would not worry much about the analytics package as analytics are not provided as part of the server anymore. I’d put more focus on the ‘Push’ where I see relevant the split core/push is for the distinction mobile/desktop/server. On the server you won’t want to store installation objects automatically, so It makes sense to not have it by default. On other platforms, you want that process to be seamless, therefore adding the package ParsePush makes this feature easy to implement.

**Comment by montymxb** [comment]
@baywet very helpful! Thank you for the input.

Sorry I have been rather absent, been incredibly busy lately, but I have been watching!

@TheFanatr thanks for putting together an updated storage dep. A potential PR would be fantastic as well.

As for versioning on the next release it would definitely be a major version change. That allows us the freedom to make fundamental changes to the SDK as is needed. Analytics has been pretty much phased out server side as flovilmart mentioned above, but not to the point where I would pull it out yet.

**Comment by baywet** [comment]
@montymxb no worries, feel free to add me as a reviewer when the PR comes in if you want me to provide some more input. I'll be out with no internet access second half of January but if work is not finished by then, happy to provide feedback.

**Comment by TheFanatr** [comment]
Ok, so does that mean I should completely scrap analytics, or should I just provide it separately?

**Comment by TheFanatr** [comment]
Uhh, so Json.Encode trips over this
`IDictionary<string, object> expected = new Dictionary<string, object> {
        {
          "data", new Dictionary<string, object> {{
            "alert", "Some Alert"
          }}
        },
        {
          "where", new Dictionary<string, object> {{
            "channels", new Dictionary<string, object> {{
              "$in", new List<string> {{ "channel" }}
            }}
          }}
        }
      };`
Should I be concerned?

**Comment by TheFanatr** [comment]
The issue is that the List's content is declared with encapsulating item curly brackets around `"channel"`, and Encode isn't recognizing the type `{ "channel" }` as anything.

**Comment by TheFanatr** [comment]
Can we switch to Newtonsoft.Json? I ended up having to use it for the unit tests. I can also now confirm that all 188 of them are successful.

**Comment by TheFanatr** [comment]
For versioning purposes, can I assume `Parse v2.0.0-beta1` as the version to use for the assemblies?

# Proposed Next Steps
- Remodularize the Project (differently)
- Remove Useless Elements
- Add Unity Support
- Add New API (vNext)

## 1. The Proposed New Structure
### `Parse [NuGet Only]`
- All of the Below

### `Parse.Common`🔗`StandardStorage`+`Newtonsoft.Json`
- Initialization
- Common Classes

### `Parse.Core`🔗`Parse.Common`+`WebSocketSharpStandard [UNRELEASED]`
- Querying [Including ParseLiveQuery Support]
- Class Management
- Object Manipulation

### `Parse.Cloud`🔗`Parse.Common`
- Cloud Coding
- Database Manipulation

### `Parse.Events`🔗`Parse.Common`+`Rx.NET`
- Event Handling for Push
- Push Sending
- Push Management

### `Parse.Telemetry`🔗`Parse.Common`
- Legacy Analytics
- Object-Based Analytics and Telemetry

Basically, the point is that each section of the API should have it's own respective assembly and package, and one should never reference another, except `Parse.Common`; `Parse` is exempt from this rule because it is NuGet only, as in there won't actually be a project representing it, rather it will be a package that includes all the rest of them. 

Please let me know what you think.

## 2. Remove Useless Elements
This one ties into #1; since the structure is being reformatted, why not get rid of useless elements such as the single-use interfaces, and custom implementations of already existing APIs included in .NET Standard 2 or available through NuGet.

## 3. Add Unity Support
This should not prove too difficult given that Unity now supports .NET Framework 4.6 experimentally, and the team is supposedly planning to implement .NET Standard 2 support; however, there was no time frame given last time I checked, other than "next major release". If support doesn't roll out soon, then it would be possible to implement support for Unity through prepocessor directives and a backwards-compatibility project.

## 4. Add New API
I would like to suggest some new bits of API to add to the project, but I do know that going to far past what other Parse platform APIs have done may not sit well; however, I do not know how the system works, so if it is ok, I would probably suggest some more. Here is what I have come up with:

### `Parse.`~~`Push`~~`Events`🔗`Parse.Common`+`Rx.NET`
- Event Handling for Broadcasting along with Pushes and Updates
- Event Management

### `Parse.Telemetry`🔗`Parse.Common`
- Activity and Error Logging

Please let me know what you guys think, and if you have any suggestions. the PR will go up sometime today.

**Comment by baywet** [comment]
@TheFanatr do you have a lot of cases where people use, let's say Parse.Common and Parse.Events but not Parse.Cloud? In that case why having a meta package "Parse" that will refer all of these? It might confuse people more than anything else and "force" them to download the whole thing when they only need some parts of it.
IMHO Parse.Common in your plan should be Parse, no need for a meta package. Dependency management will take care of it. "if I want Parse.Cloud, it's going to add Parse as well"

**Comment by flovilmart** [comment]
@TheFanatr this may be be of interest: https://blogs.unity3d.com/2018/01/05/discontinuing-support-for-monodevelop-unity-starting-in-unity-2018-1/

**Comment by TheFanatr** [comment]
@flovilmart Those are some good points; I think maybe I should just drop the additional API plans for now. As for using Rx.NET, I agree that sticking to async would probably be best for now, but the reason I wanted to use it for event management is because it is really efficient at doing a repetitive thing in the background without hogging resources. I did a profiling experiment once, and versus the `System.Threading.Timer` and starting a new thread with a for loop, it used about half the memory, and it rarely ever increased it's usage.

**Comment by TheFanatr** [comment]
@montymxb Can you please upgrade Appveyor to support VS2017? 
https://ci.appveyor.com/project/montymxb/parse-sdk-dotnet-gq28w/build/1.0.8/messages
Basically, as far as I can tell, the build is failing for #284 because it's not detecting the projects as valid due to the fact that it's trying to recognize them with a VS2015 point of view. Apparently other people have had the same problem with VSTS and all they had to do is change a few settings to enable support. Would this be possible?

**Comment by montymxb** [comment]
@TheFanatr Sorry I've been _super_ busy lately, plenty of notifications to go through here. Looks like all we would need to change for appveyor is the image referenced in the [appveyor config file](https://github.com/parse-community/Parse-SDK-dotNET/blob/master/.appveyor.yml#L2). In your PR you can update the **.appveyor.yml** file to reference **Visual Studio 2017**, instead of the current **Visual Studio 2015**. That should take care of your problem there 👍 .

**Comment by TheFanatr** [comment]
Ok, thanks. I'll push sometime soon.

**Comment by TheFanatr** [comment]
@montymxb CI seems to not be working properly. I got it to successfully build and generate a report, but it would seem that either the tests aren't running, or the generated file is not being analyzed. #284 https://codecov.io/gh/parse-community/Parse-SDK-dotNET/commit/51f91d914448c7bb26a374279df8bfbc1fb3089d

**Comment by montymxb** [comment]
Noticed those failed appveyor builds. I can't look into this atm, but I have flagged this as something I will look at within the next couple days. Most likely on Thursday when I have free time.

**Comment by TheFanatr** [comment]
So, basically, #284 works and is fully set up with code coverage and continuous integration, as far as I can tell. 
Here is what I plan to do next:
 - Remove redundant interfaces with only one implementation. These were originally used when there were multiple projects for every single Parse SDK section.
 - Rename some of the properties in various places to make more sense. (more details soon)
 - Remove AssemblyLister and all usage of it because it causes problems with .NET Core, and because it is likely no longer needed.
 - Add Unity support back, through the use of the Roslyn preprocessor and an extra project targeting .NET Framework 4.6.1, with all the files linked for compile.
 - Add more unit tests to improve coverage.
Please bring up any comments or concerns with any of these points here, as well as any suggestions. If an issue with my pull request repo is found, please post findings as a comment on the pull request itself (#284).

**Comment by montymxb** [comment]
Awesome! I just left a note that I'll be looking back at that shortly. Before we do anything I'm would like to fork the existing state of this repo into a **1.8.0-_ish_** branch. From there we'll have a stable base we can go back to in case things go haywire, but I expect it should be decent.

Once I've forked the 'stable' (if you can call it that as is...) state of the current repo we can bring changes into **master** and move forwards toward a fresh **2.0.0-alpha**. We can do w/e we want from that point until we've ascertained it's stability. Once ready we can jump to a  **2.0.0-beta** or skip to production at **2.0.0** flat. Following that release pattern we'll have freedom to fundamentally restructure as needed, and to drop/add support for things as we deem fit.

As for the things you're planning

- Cleaning up redundancy is always a +
- Renaming will depend, if it makes sense
- AssemblyLister is causing issues there? Hmm, once we get the PR in you should elaborate more so we can proceed on that.
- Unity down the road sounds good
- Yes, we need more unit tests before we start making too many changes from the structure. Tightly defining that will make it much easier to maintain down the road. This is a **priority** since the project is community run now.

@TheFanatr Also, many thanks for putting time into this. It goes without saying that we're very happy to see this moving along. Not exactly many .NET experts among the current community, so it's always good to have someone lend a hand (or a bulldozer and a complete restructure 😆 ). Anyways thanks again. I'll be back over in the PR once I've cleared out all these notifications that have been waiting for the past couple weeks.

**Comment by JonMcPherson** [comment]
Any progress with this? Development on this project seems to have died.
I am working on integrating this with a Unity project using .NET Standard 2.0 and IL2CPP.  From what I can tell so far, the PCL Storage "StandardStorage" will not work with Unity as-is, and will instead need to use the Unity API to get a safe platform dependent filesystem location using `Application.persistentDataPath`
https://docs.unity3d.com/ScriptReference/Application-persistentDataPath.html

So far, this has been the only issue I've run into, and a lot of the core Parse functionality including my LiveQuery implementation works well in Unity.

**Comment by flovilmart** [comment]
@JonMcPherson we’re looking for a maintainer on this project to be honest. None of us, at the time can provide extensive dotNet support.

**Comment by thibautvdu** [comment]
Hello everyone

I assume there are no input on this ?

The new standard for Xamarin cross-platform projects is now '.NET Standard', which makes the Nuget package (and last release's dll) incompatible. I've been struggling to find a solution.

Anyone using Parse with Xamarin here ?

**Comment by Digi59404** [comment]
Hey @thibautvdu - I’m using this on Xamarin. I’ve just added the dll as a reference in my forms project. I switch between iOS and Android dlls depending on the build.

**Comment by thibautvdu** [comment]
@Digi59404 I'm really grateful for this tip. I just tried and that does the job. I was referencing those dlls directly from the Droid and iOS projects.

**Comment by TobiasPott** [comment]
Although this issue was started for a different version of the SDK the support of .NET standard is available now and the SDK is usable on various platforms relying on .NET Standard.
